### ISSUE TITLE
Training Problem

### ISSUE THREAD

**Issue by burakceng** [andy-yun/pytorch-0.4-yolov3#48]
Hello sir,

First of all thank you for your efforts. I'd like to know whether I've been doing something wrong with training yolo with a simple small dataset (I will try to train it on a bigger dataset). Upon the tutorial mentioned in the link https://medium.com/@manivannan_data/how-to-train-yolov2-to-detect-custom-objects-9010df784f36 , I decided to get the hang of how to train yolo with a dataset that has only 1 class with 232 images. My computer does not have any GPU that can utilize CUDA; therefore, I wanted to see how the training goes with CPU and to do that I changed the function in **train.py** called _curmodel_ to return the global model only (I tried to print the number of gpus that the code detects, even though I do not own an nvidia gpu, it says I have 4, that's why I changed that function). Then, I adjusted the required .data, .names files and set the tiny_yolo.cfg to have 8 batches so that I can see a couple training steps more quickly, and since there is only 1 class, I also set the number of filters at the last convolutional layer to (1 + 5)*5 = 30 with 1 class to detect at region layer. 
As for my problem, the image I added to this issue reprsents it. The loss does not change at all (or I might be mistaken in interpretation). The situation is similar with pretrained weights as well, with a slight change in the loss values only.
![yolo_train_problem](https://user-images.githubusercontent.com/13833523/51732473-719c6d80-208f-11e9-9f55-448b90258eeb.png)
I have a couple questions regarding my situation:
1. What might I be doing wrong?
2. Is is because I do not use any GPU (when I train with my actual dataset, the machines that I'll use will have GPU's)?
3. Or am I impatient?

Thanks in advance,

Burak

EDIT: When I set the learning rate to small numbers like 1e-5 or 1e-7, the situation does not change; however, when I set it to 1, it started from losses around 0.035 and almost always decreased and I cut it with Ctrl-C when loss was around 0.009.

**Comment by burakceng** [comment]
Well, I switched to a remote gpu cluster for training and it shows more meaningful numbers, but still the class output is 0.0 almost always I guess. My questions remain still.
![new_train](https://user-images.githubusercontent.com/13833523/51786032-d62ef980-216f-11e9-8762-dae2eade0b15.png)

EDIT:
As for class loss, I guess since the examplary dataset I use has only 1 class, it will always show 0.0. 
Plus, I did not know using a GPU would make such difference compared to CPU based training.

**Comment by burakceng** [comment]
Sorry, to reopen this issue, but I wanted to ask the same question. I switched to train my model (a custom one) on a different dataset with tool images without pre-trained weights as the model is smaller compared to yolo9000 version. What is the problem with the following (other than the fact that my model does not learn at all):
![train_problem2](https://user-images.githubusercontent.com/13833523/51903894-5a7eb800-23ce-11e9-8222-c19c74d80990.png)
And
![train_problem3](https://user-images.githubusercontent.com/13833523/51903916-6a969780-23ce-11e9-9cd5-80e55336d90a.png)

When I increase the learning rate, it does not affect at all. Could it be because of the data (the dataset is small but composed of ImageNet tool images - 3 classes: hammer, drill, screwdriver - in total of 519 images)? Or about the other parameters? When I train with another dataset with only 1 class and about 230-250 images, there were not any of these issues.

EDIT: When I set the learning rate to 1 with 8 batches, 4 subdivisions, it stops at this point:
![train_problem4](https://user-images.githubusercontent.com/13833523/51905393-31f8bd00-23d2-11e9-8dbb-77a5d7c493a2.png)

**Comment by andy-yun** [comment]
If you want to cpu model, just assign use_cuda=False in train.py.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by andy-yun** [comment]
In your screenshot, the detect layer’s number is too small. Did you change the model structure? 
When you train the model, nPP should have non zero value if the training is correctly performing.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by burakceng** [comment]
Yes, I am using 2 different cfg's, one is almost the same as tiny-yolo with a few hyperparameter differences and the other is quite different from any yolo version (just going crazy with the architecture). As you have stated in another issue, I recuded the threshold from 0.5-0.6 to 0.25 and nPP increased; however, precision and recall statistics remained same, i.e. allways 0.
![tp5](https://user-images.githubusercontent.com/13833523/51986709-e61d4500-24b1-11e9-8f00-ec3a6cc6fc95.png)
![tp6](https://user-images.githubusercontent.com/13833523/51986716-eae1f900-24b1-11e9-8ce2-12700886602c.png)

When I use the tiny-yolo-like model, the training stops at 4890th iteration/batch which is right before calculating the precision and recall values for the first time, and it cannot evaluate them either, but runs indefinitely.

**Comment by andy-yun** [comment]
With your last condition, learning rate = 1 and 8 batch, 4 subdivisions, you may train the model with patience. You can skip the intermittent evaluation by -n flag because test() function seems like stopped condition along to too many bounding boxes.

**Comment by burakceng** [comment]
When I trained my models on Google Colab, they actually needed feasible learning rates like 1e-3, 1e-4 etc; otherwise, the loss exploded. When I left it in training for 3-4 hours at colab, it reached batch number ~454k. When I checked the precision and recall logs, all ~900 epochs had precision and recall and fscore 0. Even though I tweaked the proposal thresholds in order to get more nPP, it did not change anything. The bbox loss was not being updated at all, the average of it probably changed about 5 pts during that 450k+ iterations, the class loss had a same situation with having different values but changing even less than bbox loss. The total loss was almost totally based on the confidence loss, because those 2 were the only ones getting updated visibly over time. Should I increase the proposal threshold? Because after having been training for some time nPP starts to increase.

What could be the reason? Do I need more data? (I was just trying to fine-tune tiny-yolo to my dataset, which is quite small with 3 classes and 544 images in total).

Kind regards,

**Comment by andy-yun** [comment]
The threshold for nPP is not related to precision and recall percent. I recommend you change steps and scales condition. Both variables are used for adaptive learning rate or modifying learning rate along to training iteration. Among various combinations of learning_rate, steps, and scales etc., you could obtain the model fit your data.

**Comment by burakceng** [comment]
Hello again, I have a few questions:
* What about the saturation and hue values when defining a network (at the beginning of the cfg files)? I have seen the dataset generating code, I think they are used to change the images a bit so that the images with bad illumination will be more observable by the network. What if I changed those values?
* Would there be a problem if the images in the dataset were different in size? Even though the bounding box labels are scaled with respect to the image they belong to, the class instances (i.e. objects bounded by the boxes) will have different sizes, physically.
* Or maybe they are identical, but the size of input images is different?

Thanks...

**Comment by andy-yun** [comment]
@burakceng 
1) The model already did handle like your mentions. In the cfg file, you can find the saturation and exposure, hue values at the [net] session. 
2) Nope. In the training stage, various image size is considered.
3) bounding boxes of images are normalized by the image size. and the model predicts the bounding boxes in the range of 0 and 1. Lastly, the bbs are rescaled by the original image. thus, different image size is not a problem.
